servant-multipart-api/src/Servant/Multipart/API.hs

```diff
@@ -0,0 +1,256 @@
+  , fdPayload   :: MultipartResult tag
+                            -- ^ path to the temporary file that has the
+                            --   content of the user's original file. Only
+                            --   valid during the execution of your handler as
+                            --   it gets removed right after, which means you
+                            --   really want to move or copy it in your handler.
+  }
+
+deriving instance Eq (MultipartResult tag) => Eq (FileData tag)
+deriving instance Show (MultipartResult tag) => Show (FileData tag)
+
+-- | Representation for a textual input (any @\<input\>@ type but @file@).
+--
+--   @\<input name="foo" value="bar"\ />@ would appear as @'Input' "foo" "bar"@.
+data Input = Input
+  { iName  :: Text -- ^ @name@ attribute of the input
+  , iValue :: Text -- ^ value given for that input
+  } deriving (Eq, Show)
+
+-- | 'MultipartData' is the type representing
+--   @multipart/form-data@ form inputs. Sometimes
+--   you may instead want to work with a more structured type
+--   of yours that potentially selects only a fraction of
+--   the data that was submitted, or just reshapes it to make
+--   it easier to work with. The 'FromMultipart' class is exactly
+--   what allows you to tell servant how to turn "raw" multipart
+--   data into a value of your nicer type.
+--
+--   @
+--   data User = User { username :: Text, pic :: FilePath }
+--
+--   instance FromMultipart Tmp User where
+--     fromMultipart form =
+--       User \<$\> lookupInput "username" (inputs form)
+--            \<*\> fmap fdPayload (lookupFile "pic" $ files form)
+--   @
+class FromMultipart tag a where
+  -- | Given a value of type 'MultipartData', which consists
+  --   in a list of textual inputs and another list for
+  --   files, try to extract a value of type @a@. When
+  --   extraction fails, servant errors out with status code 400.
+  fromMultipart :: MultipartData tag -> Either String a
+
+instance FromMultipart tag (MultipartData tag) where
+  fromMultipart = Right
+
+-- | Allows you to tell servant how to turn a more structured type
+--   into a 'MultipartData', which is what is actually sent by the
+--   client.
+--
+--   @
+--   data User = User { username :: Text, pic :: FilePath }
+--
+--   instance toMultipart Tmp User where
+--       toMultipart user = MultipartData [Input "username" $ username user]
+--                                        [FileData "pic"
+--                                                  (pic user)
+--                                                  "image/png"
+--                                                  (pic user)
+--                                        ]
+--   @
+class ToMultipart tag a where
+  -- | Given a value of type 'a', convert it to a
+  -- 'MultipartData'.
+  toMultipart :: a -> MultipartData tag
+
+instance ToMultipart tag (MultipartData tag) where
+  toMultipart = id
+
+class MultipartBackend tag where
```

Comment:
IMO this class, along with `resourcet` dependency, belongs in `servant-multipart` package, not the `-api` one. What do you think?

Reply:
i thought the same, but then i am clueless about this:

- class `MultipartBackend` defines type `MultipartResult`
- type `MultipartResult` is needed by type `FileData`
- type `FileData` is needed by type `MultipartData`
- type `MultipartData` is needed by classes `FromMultipart` and `ToMultipart`
- and `ToMultipart` seems like a client thing

if the last bullet point is correct then it seems like moving this all into the `servant-multipart` package is not the right thing, or is it?

Reply:
Hmm, in the current state of the PR one can't actually use client without server, since client needs `MultipartResult tag` instance to construct the input.

I propose we make `MultipartResult` a top-level _open_ type family instead of associated type and define Mem and Tmp instances in this module. Then, move this class to server. Now the client user no longer needs to import server package.

Reply:
This sounds good, but this touches the edge of my haskell knowledge. I'd love to do that, but then i need instructions in a little bit more explicit steps.

Reply:
I mean you leave in this module only this:

```haskell
type family MultipartResult (tag :: *) :: *
type instance MultipartResult Mem = ...
type instance MultipartResult Tmp = ...
```

The rest goes to the respective server and client packages.

If you have no luck with this, I'll try to do it myself tomorrow.

Reply:
this is helpful, i will try it. won't have time for that today though.

Reply:
Ok i tried my luck. This is pretty straightforward up to the point where `loadFile` is used from the `servant-multipart-client` library but defined in the backend class which i moved to `servant-multipart`. While trying to move the `loadFile` function out of the backend class to the client, i somehow got lost in the whole idea around the type proxy mixed with the new type family.

Reply:
I've done it, please take a look

Reply:
Looks great, thanks!